### PR TITLE
Tags at panels

### LINKED ISSUE #4297: tag position inside plot borders
Some peer-reviewed journals (e.g. Oecologia) require figure panel labels to be _inside_ of plot borders (although they seem to enforce these requirements inconsistently).  It would be nice if there were an option like `theme(plot.tag.position = "topleftinner")` to place a label at the upper left corner of the inside of plot borders.  Currently, I think I have to kludge this by tweaking `theme(plot.tag = element_text(vjust = , hjust = )`, but it's inconsistent, so I have to do this tweaking for every panel individually.  Maybe there's an easier way that I don't know about?

Related to: https://github.com/thomasp85/patchwork/issues/222

``` r
library(ggplot2)
p1 <- ggplot(mtcars) + geom_point(aes(mpg, disp)) + labs(tag = "a)")
p1 + theme(plot.tag.position = c(0,1),
           plot.tag = element_text(vjust = 1.5, hjust = -2.85))
```

``` r
# left of "a" just touching plot border

p3 <- ggplot(mtcars) + geom_smooth(aes(disp, qsec)) + labs(tag = "b)")
p3 + theme(plot.tag.position = c(0,1),
           plot.tag = element_text(vjust = 1.5, hjust = -2.85))
#> `geom_smooth()` using method = 'loess' and formula 'y ~ x'
```

``` r
# "b)" is slightly further in with same theme()
```

<sup>Created on 2020-12-15 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### PR DESCRIPTION
**Pull request by teunbrand**
This PR aims to fix #4297.

Briefly, it adds the option to use `plot.tag.position = c({position}, "panel")` and draw the plot tag in the panel. It should also work with the patchwork package.

``` r
devtools::load_all("~/packages/ggplot2/")
#> ℹ Loading ggplot2

p <- ggplot(mpg, aes(displ, hwy)) +
  geom_point()

p +
  labs(tag = "A)") +
  theme(plot.tag.position = c("panel", "topright"))
```

``` r

library(patchwork)

(p + p) + plot_annotation(tag_levels = 'A') &
  theme(plot.tag.position = "panel")
```

<sup>Created on 2023-01-30 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Some side-notes:

* I separated out the tag logic from the already enormous `ggplot_gtable.ggplot_built()` method. I refactored a little bit along the way.
* Instead of adding yet another theme argument, I expanded the use of the already existing one. I think a lot in the theme, related to positions and alignments, is named confusingly, so I didn't want to pile on this already dizzying naming scheme (also nagged about this [here](https://github.com/tidyverse/ggplot2/issues/4020#issuecomment-1386109142)).
* I once again ran into [this bug](https://github.com/tidyverse/ggplot2/pull/4879#issuecomment-1369235237), so I fixed it in the same way.

### CODE REVIEW

File: R/plot-build.r [hunk 1]
```diff
@@ -432,3 +359,88 @@ by_layer <- function(f, layers, data, step = NULL) {
   )
   out
 }
+
+# Add the tag element to the gtable
+table_add_tag <- function(table, label, theme) {
+  # Initialise the tag margins
+  table <- gtable_add_padding(table, unit(0, "pt"))
```
Comment (teunbrand) [inline R/plot-build.r:366]: Patchwork for some reason assumes this padding is present regardless of whether it was needed, so I left this in.

File: R/plot-build.r [hunk 2]
```diff
@@ -432,3 +359,88 @@ by_layer <- function(f, layers, data, step = NULL) {
...
+    )
+    place <- find_panel(table)
+  } else {
+    n_col <- ncol(table)
+    n_row <- nrow(table)
+    # Actually fill margin with relevant units
+    if (top)    table$heights <- unit.c(height, table$heights[-1])
+    if (left)   table$widths  <- unit.c(width,  table$widths[-1])
+    if (right)  table$widths  <- unit.c(table$widths[-n_col],  width)
+    if (bottom) table$heights <- unit.c(table$heights[-n_row], height)
```
Comment (teunbrand) [inline R/plot-build.r:432]: I think this pattern was some workaround for grid units not handling subassignment in R3.2 and earlier. Can we subassign units now?

**Comment by teunbrand** [comment]
Should merge #5175 before this. Then update this PR to drop changes introduced in [5dd6c2a](https://github.com/tidyverse/ggplot2/pull/5167/commits/5dd6c2a1ee5edc324c73980fb0995f6cb551a0c3).

File: R/plot-build.r (hunk 1 above)
Comment (thomasp85) [reply R/plot-build.r:366]: yes, thanks... We should generally try to be predictable wrt the gtable structure that ggplot2 produces as it makes aligning multiple plots easier for patchwork

Comment (teunbrand) [reply R/plot-build.r:366]: That makes good sense to me :)

**Comment by thomasp85** [comment]
With this PR there will be 3 different placement strategies for tags. Either in the margin (adding space for it), on top of the full plot area (not adding space), or in the panel area (not adding space). Do all three have a name in this new API?

**Comment by teunbrand** [comment]
I also found that if you use the margin placement, `hjust/vjust` is ignored. This is isn't as much as issue if you use the plot placement, as the two numbers you give essentially determine the `hjust/vjust`.

To illustrate:

``` r
library(ggplot2)

p <- ggplot(mpg, aes(displ, hwy)) +
  geom_point() +
  labs(tag = "Foobar")
```

In the plot below, I'd expect the tag to be in the top left, with a left margin for the plot, but not a top margin for the plot.

```
p + theme(plot.tag.position = "left",
          plot.tag = element_text(vjust = 1))
```



Likewise, here I would have expected the text to be at the top right, with a plot margin on top but not on the right.

``` r
p + theme(plot.tag.position = "top",
          plot.tag = element_text(hjust = 1))
```



<sup>Created on 2023-03-21 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by thomasp85** [comment]
I like the "margin" name and think we should include that for completeness...

As for alignment, it works "correctly" right now in that it is perfectly aligned to the textbook that closely wraps the tag text. It is not given that the alignment is relative to the panel size.

I can see the argument for having it as you describe because the alignment is right now useless but I don't think it is obviously a bug as it is

File: NEWS.md [hunk 3]
```diff
@@ -19,6 +19,8 @@
   `panel.grid.major`'s linewidth setting (@teunbrand, #5179)
 * The `datetime_scale()` scale constructor is now exported for use in extension
   packages (@teunbrand, #4701).
+* The `plot.tag.position` in `theme()` can now accept `"panel"` to set the tag
+  relative to the panel area instead of the plot area (#4297).
```
Comment (thomasp85) [inline NEWS.md:23]: Can you update the bullet to include the full set of new possibilities

**Comment by thomasp85** [comment]
Can we somehow make this new alignment work with numeric positions..?

**Comment by teunbrand** [comment]
We could in theory, but the way it is setup right now it isn't straightforward, as we cannot pass a mixed type character/numeric to the `plot.tag.position` setting. There is currently *some* flexibility of placement, in that you can set the hjust/vjust if only one of the directions is fixed:

``` r
devtools::load_all("~/packages/ggplot2/")
#> ℹ Loading ggplot2

p <- ggplot(mpg, aes(displ, hwy)) +
  geom_point() +
  labs(tag = "Foobar")

p + theme(plot.tag.position = c("panel", "top"),
          plot.tag = element_text(hjust = 0.75))
```



<sup>Created on 2023-04-15 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

I think we might be able to extend this by not having the position default to 'topleft' for the 'panel' setting.

**Comment by thomasp85** [comment]
yeah, I know the limitations of the current setup. I was thinking we should maybe considering a different setup such as providing an additional theme element (e.g. `plot.tag.location`) that holds the `margin`/`plot`/`panel` info... this would also bring it somewhat more in line with other location styles

**Comment by teunbrand** [comment]
We could make that work, but how should a numeric value interact with `plot.tag.location = "margin"`?

**Comment by thomasp85** [comment]
yeah, that's a good question... 

But I think having that not work (and documented) is better than the alternative

**Comment by teunbrand** [comment]
I'm just throwing an error now with numeric `plot.tag.position` and `plot.tag.location = "margin"`.

**Review by thomasp85 — APPROVED** [review]
LGTM

**Pull request merged**